src/mlpack/core/optimizers/spalera_sgd/spalera_sgd_impl.hpp

```diff
@@ -0,0 +1,208 @@
+  for (size_t i = 1; i != maxIterations; ++i, ++currentBatch)
+  {
+    // Is this iteration the start of a sequence?
+    if ((currentBatch % numBatches) == 0)
+    {
+      // Output current objective function.
+      Log::Info << "SPALeRA SGD: iteration " << i << ", objective "
+          << overallObjective << "." << std::endl;
+
+      if (std::isnan(overallObjective) || std::isinf(overallObjective))
+      {
+        Log::Warn << "SPALeRA SGD: converged to " << overallObjective
+            << "; terminating with failure.  Try a smaller step size?"
+            << std::endl;
+        return overallObjective;
+      }
+
+      if (std::abs(lastObjective - overallObjective) < tolerance)
+      {
+        Log::Info << "SPALeRA SGD: minimized within tolerance " << tolerance
+            << "; terminating optimization." << std::endl;
+        return overallObjective;
+      }
+
+      // Reset the counter variables.
+      lastObjective = overallObjective;
+      overallObjective = 0;
+      currentBatch = 0;
+
+      if (shuffle)
+        visitationOrder = arma::shuffle(visitationOrder);
+    }
+
+    // Evaluate the gradient for this mini-batch.
+    const size_t offset = batchSize * visitationOrder[currentBatch];
+    function.Gradient(iterate, offset, gradient);
+    if (visitationOrder[currentBatch] != numBatches - 1)
+    {
+      for (size_t j = 1; j < batchSize; ++j)
+      {
+        arma::mat funcGradient;
+        function.Gradient(iterate, offset + j, funcGradient);
+        gradient += funcGradient;
+      }
+
+      // Now update the iterate.
+      if (!updatePolicy.Update(stepSize, currentObjective, batchSize,
+          numFunctions, iterate, gradient))
+      {
+          Log::Warn << "SPALeRA SGD: converged to " << overallObjective << "; "
+              << "terminating with failure.  Try a smaller step size?"
+              << std::endl;
+          return overallObjective;
+      }
+
+      // Add that to the overall objective function.
+      currentObjective = function.Evaluate(iterate, offset);
+      for (size_t j = 1; j < batchSize; ++j)
+        currentObjective += function.Evaluate(iterate, offset + j);
+
+      overallObjective += currentObjective;
+      currentObjective /= batchSize;
+    }
+    else
+    {
+      // Handle last batch differently: it's not a full-size batch.
+      const size_t lastBatchSize = numFunctions - offset - 1;
+      for (size_t j = 1; j < lastBatchSize; ++j)
+      {
+        arma::mat funcGradient;
+        function.Gradient(iterate, offset + j, funcGradient);
+        gradient += funcGradient;
+      }
+
+      // Ensure the last batch size isn't zero, to avoid division by zero before
+      // updating.
+      bool learningRateCheck = true;
+      if (lastBatchSize > 0)
+      {
+        // Now update the iterate.
+        learningRateCheck = updatePolicy.Update(stepSize, currentObjective,
+            lastBatchSize, numFunctions, iterate, gradient);
```

Comment:
If I am not mistaken, everything "special" about SPALeRA is contained in the `SPALeRAStepSize` class.  So could we refactor `SGD` to add a `DecayPolicy` type and then use `SGD<VanillaUpdate, SPALeRAStepSize>` as the SPALeRA optimizer?

Reply:
I thought about that, and discarded that idea, because the current policies just take ``iterate``, ``stepSize`` and ``gradient`` as parameter.